### ISSUE TITLE
Implement FIFO buffer for preference comparison labels

### ISSUE THREAD

**Issue by ejnnr** [HumanCompatibleAI/imitation#389]
We currently maintain a dataset of preference comparison labels that can grow arbitrarily large as we continue collecting preference comparisons during training. This has two issues:
1. Since we always train the reward model over the entire dataset in each iteration, the reward model is effectively trained more later on (which is perhaps even the opposite of what we want). This would be easy to fix though by fixing the number of steps, instead of epochs.
2. More importantly, there may be significant distributional shift because of the changing policy, and once the dataset has grown very large, new labels will have little effect---we're effectively training on an outdated transition distribution.

Because of this second point, the [DRLHP paper](https://arxiv.org/pdf/1706.03741.pdf) uses a buffer of 3,000 preference comparisons (at least for Atari). I think this might help significantly with performance in some cases, so we should implement it as well.

This means the `PreferenceDataset` class needs to be generalized so that it can take a `max_size` argument or similar, specifying the maximum allowed number of comparisons to store. If `max_size=None`, it should revert to the current behavior of growing without limit. Otherwise, it should behave as a FIFO buffer (i.e. discard the oldest comparisons when new ones are added via `PreferenceDataset.push()`).

Since the `PreferenceDataset` instance is created inside `PreferenceComparisons.__init__()`, the `max_size` argument needs to be added there as well and passed on (and it should be added to the training script).

The simplest implementation would just continue using lists to store everything (overhead from copying when we need to drop prefixes of the list is probably not too big). Pythons `collections.deque` has O(n) random access, so I'd guess it's less efficient (we need random access during reward model training).

**Comment by kmdanielduan** [comment]
Yeah, it makes sense to implement a more generalised buffer for `PreferenceDataset`. 

> 1. Since we always train the reward model over the entire dataset in each iteration, the reward model is effectively trained more later on (which is perhaps even the opposite of what we want). This would be easy to fix though by fixing the number of steps, instead of epochs.

This makes sense, but I suppose getting human feedback is perhaps considered more costly than reward training on increasingly larger dataset. Intuitively, people might want to make the best of all the human preferences collected so far rather than only the ones being collected most recently. There's also a question of how we'll use the early comparisons.

> 2. More importantly, there may be significant distributional shift because of the changing policy, and once the dataset has grown very large, new labels will have little effect---we're effectively training on an outdated transition distribution.

It's true that the reward function will be trained on early comparisons for a lot more iterations accumulatively, and new labels might have little effect. It seems to me the comparison dataset distribution will still shift if we keep throwing away early comparisons. There is also a question of whether there might be catastrophic forgetting in reward models.

Overall, I think it's worth implementing this feature, even for some ablation experiments. Happy to contribute.

**Comment by AdamGleave** [comment]
An alternative would be to keep all the preferences collected around, but weight new ones more. If we know the total number of epochs $T$ we're going to train for, I think we could rig the weights $w_{i,t}$ for data point i training at epoch t first collected at epoch $t'$  such that $\sum_{t=t'}^T w_{i,t}$ is a constant.
